### DISCUSSION TITLE
Is it possible to support index setter through parentheses instead of brackets in V8Engine?

### DISCUSSION THREAD

**Discussion by ZelgiusLiu** [ClearFoundry/ClearScript#520] [Q&A]
Hi, 
I am now trying to upgrade some old projects using JScriptCodeProvider to newer version of .Net.
And when converting the JScrriptCodeProvider to ClearScript, I found some interesting differences.
As shown in the following code, I found that for JScript, it supports both getting and setting operations with index operator through parentheses; then, for V8Engine,  it supports only getting value through parentheses, and setting operation can only be done when using brackets:
```CS
  var values = new PropertyBag()
  {
      ["greeting"] = "Hello",
      ["greeting2"] = "Hi"
  };

  using var js = new JScriptEngine();
  js.AddHostType(typeof(Console));
  js.AddHostObject("V", values);

  js.Execute("Console.WriteLine(V('greeting'))"); // Output: Hello
  js.Execute("V('greeting') = V('greeting2')"); // Works fine
  
  using var v8=new V8ScriptEngine();
  v8.AddHostType(typeof(Console));
  v8.AddHostObject("V", values);

  v8.Execute("Console.WriteLine(V('greeting'))"); // Output: Hi
  v8.Execute("V['greeting'] = 'Hello'"); // Works fine
  v8.Execute("V('greeting') = 'Hello'"); //ReferenceError: Invalid left-hand side in assignment
```
Here is what confuses me:
1. Why can we use parentheses as index operator?
2. Is it possible to support using parentheses as index operator, especially as setter, in V8Engine?

**Comment by ClearScriptLib** [comment]
Hi @ZelgiusLiu,

>Why can we use parentheses as index operator?

It's a JScript-specific language extension that facilitates host object access.

JScript and VBScript interact with host objects via [`IDispatch`](https://learn.microsoft.com/en-us/windows/win32/api/oaidl/nn-oaidl-idispatch), an interface that supports parameterized properties and other features not found in JavaScript. JScript's designers apparently felt that parameterized properties were important enough to warrant a JavaScript language extension.

>Is it possible to support using parentheses as index operator, especially as setter, in V8Engine?

V8 supports only standard JavaScript language features. It's an open-source project, so modifications are certainly possible, but this sort of change is outside the scope of the ClearScript project.

.NET does have support for parameterized properties, although in C# it's limited to indexers. ClearScript allows standard JavaScript code to access such properties via the following somewhat clunky pattern:

``` JavaScript
value = obj.Item.get(arg1, arg2, ...); // OR: value = obj.Item(arg1, arg2, ...);
obj.Item.set(arg1, arg2, ..., value);
```

Here "Item" is the name of the property. Most C# indexers are actually named "Item" under the covers, but other languages may use different names.

Good luck!
